### ISSUE TITLE
Improve stacktrace

### ISSUE THREAD

**Issue by Conaclos** [bearror/oletus#21]
Hi!

I have often problem to locate which assertion fails in my unit tests. Let's just take an example to illustrate this.

Usually my unit tests have several assertions, such as:

```js
import test from "oletus"
test("example", t => {
    t.ok(true)
    t.ok(false)
})
```

The reported location is a bit evasive:

```sh
$ oletus test.js
x error
test.js:2
```

We don't know whether it is the first or the second assertion that failed. It could be nice to get the line where the assertion failed:

```sh
$ oletus test.js
x error
test.js:4
```

However, I see the big picture: we want to support any test pattern and assertion library. It is not so easy to report a helpful location. The current reporting is evasive but reliable.

Maybe we could report two locations (a reliable one: the actual one, and an unreliable one that helps in most of usual cases):

```sh
$ oletus test.js
x example
test.js:2
test.js:4
```

The first line is still extracted with `getLocation`. The second one would be generated by a "hacky" extractor (similar to `extractTraceFallback`). It just extracts the first frame.

**Comment by bearror** [comment]
Heya! I see the point, and it'd definitely be a worthwhile change to output the most helpful location; the actual failure.

While Oletus currently does inject the `assert.strict` module in a hard-coded manner _(that was a pretty poor call from me back in the day 😅)_, you're right in that it should make as few assumptions about the assertions and testing patterns as possible. That is, you should be able to bring whatever tools suit you best, since Oletus only aims to be a test runner.

With that said, I don't actually know if there is any issue in getting the location of the top-level error from the stack trace. Conceptually — regardless of the mechanism you use to assert — you most likely end up throwing an error in the scope of the test implementation, right? If the location of that can be wrangled from the stack trace, then I think that would be just fine as a replacement of the currently reported test location.

Have to admit that I'm way rusty when it comes to Node stack traces, though, so I'll have to do some digging before I'll be able to see what it would imply — or even if there is some obvious technical hurdle. If you know of some use-case where this kind of reporting would be unreliable _(or even impossible)_, I'd be very interested to hear!

**Comment by Conaclos** [comment]
> you most likely end up throwing an error in the scope of the test implementation, right? If the location of that can be wrangled from the stack trace, then I think that would be just fine as a replacement of the currently reported test location.

Not necessarily: an implementation can delegate the test to a generic implementation such as:

```js
import { default as test } from "oletus"
// this function can be used by several test implementations
function genericTest(t, val) {
    t.ok(val)
}

test("complex", (t) => {
    genericTest(t, false)
})

test("simple", (t) => {
    t.ok(false)
})
```
The current output is:

```sh
✖ complex
tests/a.test.js:7
false == true

✖ simple
tests/a.test.js:11
false == true
```

I realized that in the complex case we need more than a single location. We could print a sanitized stack trace. Most of the time (the simple case) we should have a single line (the line of the failing assertion). This could looks like:

```sh
✖ complex
tests/a.test.js:4
tests/a.test.js:8
false == true

✖ simple
tests/a.test.js:12
false == true
```

I see a single drawback of omitting the test location. If we have implementation factories, we lose the test location:

```js
import { default as test } from "oletus"

function testFactory() {
    return (t) => {
        t.ok(false)
    }
}

test("complex", testFactory())
```

We would obtain:

```sh
✖ complex
tests/a.test.js:5
false == true
```

I think that's acceptable because most of the time test titles are unique. It seems very unlikely to meet both implementation factories and duplicated titles.

Another possibility could be to print the test location with the title:

```sh
✖ complex (tests/a.test.js:9)
tests/a.test.js:5
false == true
```
